### PR TITLE
Bump bridge darwinia-ethereum

### PR DESCRIPTION
**Pull request by fewensa**

### CODE REVIEW

**Comment by fewensa** [comment]
```
 {
          name: EthereumRelay
          storage: {
            prefix: DarwiniaEthereumRelay
            items: [
              {
                name: ConfirmedHeaderParcels
                modifier: Optional
                type: {
                  Map: {
                    hashers: [
                      Identity
                    ]
                    key: 8
                    value: 421
                  }
                }
                fallback: 0x00
                docs: [
                   Confirmed ethereum header parcel
                ]
              }
```


```
{
          name: EthereumRelayerGame
          storage: {
            prefix: Instance1DarwiniaRelayerGame
            items: [
              {
                name: RelayHeaderParcelToResolve
                modifier: Default
                type: {
                  Plain: 84
                }
                fallback: 0x00
                docs: [
                   Active games' relay header parcel's ids
                ]
              }
```



```
{
          name: EthereumRelayAuthorities
          storage: {
            prefix: Instance1DarwiniaRelayAuthorities
            items: [
              {
                name: Candidates
                modifier: Default
                type: {
                  Plain: 546
                }
                fallback: 0x00
                docs: [
                   Anyone can request to be an authority with some stake
                   Also submit your signer at the same time (for ethereum: your ethereum address in H160 format)
                  
                   Once you requested, you'll enter the candidates
                  
                   This request can be canceled at any time
                ]
              }
```

File: bridges/darwinia-ethereum/bridge/src/service/affirm/handler.rs [hunk 1]
```diff
@@ -63,26 +61,26 @@ impl AffirmHandler {
...
+            client,
             shadow,
         })
     }
 }
 
 impl AffirmHandler {
     pub async fn affirm(&mut self) -> color_eyre::Result<()> {
-        let last_confirmed = self.darwinia.last_confirmed().await?;
+        let last_confirmed = self.client.ethereum().last_confirmed().await?;
```
Comment (HackFisher) [inline bridges/darwinia-ethereum/bridge/src/service/affirm/handler.rs:77]: To double check

File: bridges/darwinia-ethereum/bridge/src/service/affirm/handler.rs [hunk 2]
```diff
@@ -159,17 +163,17 @@ impl AffirmHandler {
...
-            "Prepare to affirm ethereum block: {}",
+            "[ethereum] [affirm] Prepare to affirm ethereum block: {}",
             target
         );
 
         match self.shadow.parcel(target as usize).await {
             Ok(parcel) => {
                 let parcel: EthereumRelayHeaderParcel = parcel.try_into()?;
-                if parcel.header == EthereumHeader::default() || parcel.mmr_root == [0u8; 32] {
+                if parcel.parent_mmr_root.to_fixed_bytes() == [0u8; 32] {
```
Comment (HackFisher) [inline bridges/darwinia-ethereum/bridge/src/service/affirm/handler.rs:173]: To double check

File: bridges/darwinia-ethereum/bridge/src/service/check.rs [hunk 3]
```diff
@@ -86,29 +86,26 @@ async fn run(tracker: &Tracker) -> color_eyre::Result<()> {
...
+        let verified = match client.ethereum().is_verified(&tx_hash, tx_index).await {
             Ok(v) => v,
             Err(e) => {
-                if let Some(substrate_subxt::Error::Rpc(_)) =
-                    e.downcast_ref::<substrate_subxt::Error>()
-                {
-                    return Err(e);
-                }
-                let err_msg = format!("{:?}", e).to_lowercase();
-                if err_msg.contains("restart") {
```
Comment (HackFisher) [inline bridges/darwinia-ethereum/bridge/src/service/check.rs:106]: To double check

File: bridges/darwinia-ethereum/bridge/src/service/darwinia/scan_authorities_change_signed_event.rs [hunk 4]
```diff
@@ -16,11 +14,18 @@ impl<'a> ScanAuthoritiesChangeSignedEvent<'a> {
...
+        // let current_term: u32 = client
+        //     .runtime()
+        //     .storage()
+        //     .ethereum_relay_authorities()
+        //     .next_term(None)
+        //     .await?;
+        // fixme: It's not a good way, the best is update storage prefix with pallet name
+        let current_term = client
+            .ethereum()
+            .ethereum_relay_authorities_next_term()
```
Comment (HackFisher) [inline bridges/darwinia-ethereum/bridge/src/service/darwinia/scan_authorities_change_signed_event.rs:28]: To double check

File: bridges/darwinia-ethereum/bridge/src/service/extrinsics/handler.rs [hunk 5]
```diff
@@ -270,30 +248,21 @@ impl ExtrinsicsHandler {
...
                     Ok(_) => self.message_kv.delete(&key)?,
                     Err(err) => {
-                        if let Some(substrate_subxt::Error::Rpc(_)) = err.downcast_ref::<substrate_subxt::Error>() {
-                            tracing::warn!(
-                                target: "darwinia-ethereum",
-                                "Connection Error. Try to resend later. extrinsic: {:?}",
-                                ex,
-                            );
-                            tokio::time::sleep(std::time::Duration::from_secs(5)).await;
-                            break;
```
Comment (HackFisher) [inline bridges/darwinia-ethereum/bridge/src/service/extrinsics/handler.rs:287]: To double check.

Should try again?

File: bridges/darwinia-ethereum/bridge/src/service/guard.rs [hunk 6]
```diff
@@ -165,33 +144,40 @@ impl GuardService {
...
-                                target: "darwinia-ethereum",
-                                "The parcel of ethereum block {} from Shadow service is blank, the err msg is {}",
-                                block,
-                                msg
-                            );
-                            return Ok(extrinsics);
-                        }
-                        return Err(err);
+            if last_confirmed <= pending_block_number {
+                continue;
```
Comment (HackFisher) [inline bridges/darwinia-ethereum/bridge/src/service/guard.rs:148]:
```
if pending_block_number > last_confirmed
                && !ethereum2darwinia.has_voted(&guard_account, voting_state)
```

Why continue and not vote here, different with the condition before change.

Comment (fewensa) [reply bridges/darwinia-ethereum/bridge/src/service/guard.rs:148]: The logic here is still the same, just the order has been adjusted.

before

```
for pending in pending_headers {
  if pending_block_number > last_confirmed
                  && !ethereum2darwinia.has_voted(&guard_account, voting_state)
              {
   // do something
  }
}
```

after

```
for pending in pending_headers {
  if last_confirmed <= pending_block_number {
    continue;
  }
  let real_account = client.account().real_account();
  let has_voted = voting_state.ayes.contains(real_account)
      || voting_state.nays.contains(real_account);
  if has_voted {
      continue;
  }

  // do something
}
```

https://github.com/darwinia-network/bridger/blob/c8c1a5babff6cc3d741fa4b26ce6d5b3e39c1ac6/bridges/darwinia-ethereum/client-darwinia/src/from_ethereum/api.rs#L305
https://github.com/darwinia-network/bridger/blob/c8c1a5babff6cc3d741fa4b26ce6d5b3e39c1ac6/frame/supports/support-ethereum/src/affirmation.rs#L41

Comment (fewensa) [reply bridges/darwinia-ethereum/bridge/src/service/guard.rs:148]: I thought again. This is indeed wrong

File: bridges/darwinia-ethereum/bridge/src/service/affirm/handler.rs (hunk 1 above)
Comment (fewensa) [reply bridges/darwinia-ethereum/bridge/src/service/affirm/handler.rs:77]: before

https://github.com/darwinia-network/bridger/blob/c8c1a5babff6cc3d741fa4b26ce6d5b3e39c1ac6/bridges/darwinia-ethereum/client-darwinia/src/frame/ethereum/relay.rs#L162-L168

https://github.com/darwinia-network/bridger/blob/c8c1a5babff6cc3d741fa4b26ce6d5b3e39c1ac6/bridges/darwinia-ethereum/client-darwinia/src/from_ethereum/api.rs#L197-L206

after

https://github.com/darwinia-network/bridger/blob/2e4afda99f8bd9520f66b2714ccb1efb03e8a8f1/bridges/darwinia-ethereum/client-darwinia/src/ethereum/api.rs#L31-L41

File: bridges/darwinia-ethereum/bridge/src/service/affirm/handler.rs (hunk 2 above)
Comment (fewensa) [reply bridges/darwinia-ethereum/bridge/src/service/affirm/handler.rs:173]: Before

https://github.com/darwinia-network/bridger/blob/c8c1a5babff6cc3d741fa4b26ce6d5b3e39c1ac6/frame/supports/support-ethereum/src/parcel.rs#L17


https://github.com/darwinia-network/bridger/blob/c8c1a5babff6cc3d741fa4b26ce6d5b3e39c1ac6/frame/supports/support-ethereum/src/block.rs#L15

After

https://github.com/darwinia-network/bridger/blob/2e4afda99f8bd9520f66b2714ccb1efb03e8a8f1/frame/assistants/darwinia-subxt/src/x_shadow.rs#L18

File: bridges/darwinia-ethereum/bridge/src/service/check.rs (hunk 3 above)
Comment (fewensa) [reply bridges/darwinia-ethereum/bridge/src/service/check.rs:106]: https://github.com/darwinia-network/bridger/pull/434/commits/db46c9252044bb031fc4ccb169e7216ba12131b4

**Review by HackFisher — APPROVED** [review]

**Pull request merged**